admin/sql/updates/2021-04-12-new-music-services-table.sql

```diff
@@ -0,0 +1,25 @@
+BEGIN;
+
+CREATE TYPE external_auth_service AS ENUM ('spotify', 'youtube');
+
+CREATE TABLE external_auth (
+    id                      SERIAL,
+    user_id                 INTEGER NOT NULL,
+    service                 external_auth_service NOT NULL,
+    access_token            TEXT NOT NULL,
+    refresh_token           TEXT,
+    token_expires           TIMESTAMP WITH TIME ZONE,
+    last_updated            TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
+    record_listens          BOOLEAN NOT NULL,
+    service_details         JSONB
+);
+
+ALTER TABLE external_auth ADD CONSTRAINT external_auth_pkey PRIMARY KEY (id);
```

Comment:
I see in your original proposal in the document that you were going to use (user_id, service) as the primary key. If you still want to ensure that we can have only 1 user,service row you could add a unique index over these two fields.

We'll also need other generic indexes - I'm not 100% about the kinds of queries that you'll need to do, but it sounds like one on user_id, one on service, and one on (user_id, service) is probably useful.

Reply:
Yes, @mayhem suggested that we add a id SERIAL as well so I think it makes sense to use it as the PK. I'll add the unique index on \ (user_id, service). I think adding an index on user_id might be useful but on service might not be. My crude understanding (maybe wrong) of an index is to allow using binary search on the contents. But for service there are going to very few values and large number of occurences of the same value. The other two indices will probably cover it well enough to not create this one. We can add remaining indices later as the need arises.

Reply:
A good rule of thumb is simply "if you want to `select WHERE <this field>=something` then you should have an index".
So if you need to select all users for a service we should have an index on service. You're right that with the current setup we may have few enough rows that it doesn't make a huge difference, but imo it's always better safe than sorry here.

Reply:
Makes sense. Agreed, will add.